flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/deployment/AbstractDeploymentObserver.java

```diff
@@ -211,6 +219,22 @@ protected boolean isSuspendedJob(FlinkDeployment deployment) {
                 && lastReconciledSpec.getJob().getState() == JobState.SUSPENDED;
     }
 
+    private void onMissingDeployment(FlinkDeployment deployment) {
```

Comment:
Can we distinguish the `onMissingDeployment` cause here (like by judge whether there is an on-going upgrade). If it caused by Operator, we can continue reconcile. Otherwise we can just clean up the `FlinkDeployment` which means an external operation or internal finish/failed happened.

Reply:
On-going upgrade would never trigger this , when the job is in suspended state with missing deployment we won't try to observe it. 

Even if we trigger there is a problem we should never clean up the deployment, as the user probably needs to manually recover the job. Cleanup is reserved for deletions.